### DISCUSSION TITLE
Can't upload file from ReactNative

### DISCUSSION THREAD

**Discussion by haswalt** [urql-graphql/urql#1791] [Q&A] [answered]
In a fairly simple example I can't get uploads to work. I'm using a go server generated by gqlgen. I've tested uploads through several other clients and they work fine, but when trying in ReactNative I get an error returned. Looks like something isn't being translated correctly when setting up the multipart request.

Client:

```
const client = createClient({
    url: 'http://localhost:9080/query',
    exchanges: [multipartFetchExchange],
  });
```

Sample:

```
const UPLOAD_FILE = gql`
  mutation UpdateAvatar($file: Upload!) {
    updateAvatar(file: $file) {
      profile {
        avatar
      }
    }
  }
`;

const [result, uploadFile] = useMutation(UPLOAD_FILE)

const file = new ReactNativeFile({
  uri: {uri_from_image_library},
  type: 'image/jpg,
  name: 'avatar.jpeg',
});

uploadFile({ file })
```

When firing off this mutation I get back an error of `[CombinedError: [GraphQL] map[string]interface {} is not an Upload]`. Some debug logs:

```
LOG  {"name": "avatar.jpeg", "type": "image/jpg", "uri": "file:///Users/harry/Library/Developer/CoreSimulator/Devices/C88F4F9B-B931-4DE4-9707-2FDD2E047D2B/data/Containers/Data/Application/5F538E39-4C90-4A5C-A64D-92DAB11D2DEB/tmp/0DA5DAEA-578D-4F5D-8A09-31432BA27077.jpg"}
 LOG  multipartFetchExchange fetchRequest
 LOG  {"context": {"authAttempt": false, "fetch": undefined, "fetchOptions": {"headers": [Object]}, "preferGetMethod": false, "requestPolicy": "cache-first", "suspense": false, "url": "http://localhost:9080/query"}, "key": 1115913409, "kind": "mutation", "query": {"__key": 1672471653, "definitions": [[Object]], "kind": "Document", "loc": {"end": 105, "source": [Object], "start": 0}}, "variables": {"file": {"name": "avatar.jpeg", "type": "image/jpg", "uri": "file:///Users/harry/Library/Developer/CoreSimulator/Devices/C88F4F9B-B931-4DE4-9707-2FDD2E047D2B/data/Containers/Data/Application/5F538E39-4C90-4A5C-A64D-92DAB11D2DEB/tmp/B74B64D6-DA65-42D5-93C3-B3289230589A.jpg"}}}
 LOG  {"fetchOptions": {"body": "{\"query\":\"mutation UpdateAvatar($file: Upload!) {\\n  updateAvatar(file: $file) {\\n    profile {\\n      avatar\\n    }\\n  }\\n}\\n\",\"operationName\":\"UpdateAvatar\",\"variables\":{\"file\":{\"uri\":\"file:///Users/harry/Library/Developer/CoreSimulator/Devices/C88F4F9B-B931-4DE4-9707-2FDD2E047D2B/data/Containers/Data/Application/5F538E39-4C90-4A5C-A64D-92DAB11D2DEB/tmp/B74B64D6-DA65-42D5-93C3-B3289230589A.jpg\",\"name\":\"avatar.jpeg\",\"type\":\"image/jpg\"}}}", "headers": {"content-type": "application/json"}, "method": "POST"}, "url": "http://localhost:9080/query"}
```

Checking myself if the files are being correctly constructed they appear to be. I'm using `ReactNativeFile` from the same package as that used to check if files are added to the mutation `extract-files`.

In a slightly contrived test:

```
const {files} = extractFiles({
  variables: {
    file: file
  },
});

console.log(files.size); => 1
```

At an outside glance it looks like the content-type isn't set correctly so isn't being recognised as an upload.

On: 

* urql 2.0.3
* @urql/exchange-multipart-fetch 0.1.12

**Comment by kitten** [comment]
it's probably worth trying two things here; we haven't checked uploads on React Native for a while and it's possible that `extract-files` has pushed a fix for it. You could add a resolution for it and force upgrade it if you're using Yarn. The only thing to make sure of when you're doing that is looking at `yarn list --pattern extract-files` to ensure that the upgraded version isn't duplicated and actually used.

The other important part here is that you may want to check the multipart upload itself. As you're showing, the file is correctly identified and hence it stands to reason that the multipart request is started correctly. So, since the error you're seeing is a server-side error it may stand to reason that the server may not accept the file correctly. You could eliminate this possibility by quickly testing a web-based uploader. We have a base example for that here that should make it a breeze to quickly test this: https://github.com/FormidableLabs/urql/tree/main/examples/with-multipart

**Reply by haswalt** [reply]
Thanks Phil.

I think I can rule out the server side part. It's a fairly well tested library and other clients upload correctly, I've tested using the multipart example and the upload completes as expected.

I am using a slightly older version of extract-files which could be a cuplrit. thanks fo the heads up there, i'll upgrade and report back.

**Reply by haswalt** [reply]
Upgrading extract-files has also had no impact. As an aside i've noticed that `subscribeToDebugTarget` has stopped logging anything as well since upgrading urql from 2.0.3 to 2.0.4.

**Reply by haswalt** [reply]
Tried using @urql/devtools and I don't see any multipart-fetch exchange operations showing in their either. Is there something odd with the latest client that breaks debugging?

**Comment by haswalt** [comment, marked as answer]
@kitten you got me looking in the right place with updating `extract-files`, except it's the version in the `@urql/exchange-multipart-fetch` package that is too old and no longer compatible. I've created a PR to resolve this by upgrading the version: https://github.com/FormidableLabs/urql/pull/1795
